react/features/base/session/middleware.js

```diff
@@ -0,0 +1,164 @@
+
+        if (session) {
+            dispatch(sessionStarted(session));
+        } else {
+            console.error('CONFERENCE_JOINED - no session found');
+        }
+
+        break;
+    }
+    case CONFERENCE_WILL_JOIN: {
+        const { conference } = action;
+        const { connection } = conference;
+
+        const session = findSessionForConnection(getState(), connection);
+
+        if (session) {
+            session.conference = conference;
+        } else {
+            console.error('CONFERENCE_WILL_JOIN - no session found');
+        }
+        break;
+    }
+    case LOAD_CONFIG_ERROR: {
+        const { locationURL, room } = action;
+
+        // There won't be a session if there's no room (it happen when the config is loaded for the welcome page).
+        if (room) {
+            const session = findSessionForLocationURL(getState(), locationURL);
+
+            if (session) {
+                dispatch(sessionFailed(session));
+            } else {
+                console.error(`LOAD_CONFIG_ERROR - no session found for: ${toURLString(locationURL)}`);
+            }
+        }
+        break;
+    }
+    case CONNECTION_DISCONNECTED: {
+        const { connection } = action;
+        const session = findSessionForConnection(getState(), connection);
+
+        if (session) {
+            session.connection = null;
+            dispatch(
+                session.conferenceFailed ? sessionFailed(session) : sessionTerminated(session));
+        } else {
+            console.error('CONNECTION_DISCONNECTED - no session found');
+        }
+        break;
+    }
+    case CONNECTION_FAILED: {
+        const { connection, error } = action;
+        const { recoverable } = error;
+        const session = findSessionForConnection(getState(), connection);
+
+        if (session) {
+            session.connection = null;
+            if (typeof recoverable === 'undefined' || recoverable === false) {
+                dispatch(sessionFailed(session));
+            }
+        } else {
+            console.error('CONNECTION_FAILED - no session found');
+        }
+        break;
+    }
+    case CONNECTION_WILL_CONNECT: {
+        const { connection, locationURL } = action;
+        const session = findSessionForLocationURL(getState(), locationURL);
+
+        if (session) {
+            session.connection = connection;
+        } else {
+            console.error(`CONNECTION_WILL_CONNECT - no session found for: ${toURLString(locationURL)}`);
+        }
+        break;
+    }
+    case SET_ROOM: {
+        const { locationURL } = getState()['features/base/connection'];
+        const session = findSessionForLocationURL(getState(), locationURL);
+
+        // Web has different logic for creating the local tracks and starting the connection
+        if (session && typeof APP === 'undefined') {
+            dispatch(createDesiredLocalTracks());
+            dispatch(connect());
```

Comment:
question: should we wait until the tracks are creating before doing the connection?

Reply:
So it's for us to decide. We used to do both kind of in parallel. Maybe this way the user gets to the conference faster.

Reply:
Indeed, I had forgotten about that! Disregard.